src/v/serde/rw/variant.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/vlog.h"
+#include "serde/rw/rw.h"
+
+#include <array>
+#include <utility>
+#include <variant>
+
+// # Variant Wire Compatibility:
+//
+// Variants are tricky w.r.t. compatibility, they are forward compatible if new
+// discrimiants are appended to the variant.
+//
+// Concretely that means that you can only read old versions of a variant if new
+// options are appended. Writing a new variant option should happen with a
+// feature barrier, so that an older reader will never read a newly written
+// variant.
+namespace serde {
+
+template<typename... T>
+void tag_invoke(tag_t<write_tag>, iobuf& out, std::variant<T...> v) {
+    write<size_t>(out, v.index());
+    std::visit(
+      [&out](auto&& v) {
+          write(out, std::forward<std::decay_t<decltype(v)>>(v));
+      },
```

Comment:
```suggestion
      [&out]<typename>(T&& v) {
          write(out, std::forward<T>(v));
      },
```

or just std::move since from context this is the only option

Reply:
Nice!